File: leetcode/c++/213_house_robber_ii.cpp
// https://leetcode.cn/problems/house-robber-ii/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) {
            return nums[0];
        } else if (n == 2) {
            return max(nums[0], nums[1]);
        }
        vector<int>dp(n);
        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);
        for (int i = 2; i < n-1; i++) {
            dp[i] = max(dp[i-1], dp[i-2] + nums[i]);
        }
        int max1 = dp[n-2];
        dp[1] = nums[1];
        dp[2] = max(nums[1], nums[2]);
         for (int i = 3; i < n; i++) {
            dp[i] = max(dp[i-1], dp[i-2] + nums[i]);
        }
        int max2 = dp[n - 1];
        return max(max1, max2);
    }
};

File: leetcode/golang/213_house_robber_ii.go
func rob(nums []int) int {
    if len(nums) == 1 {
        return nums[0]
    }
    if len(nums) == 2 {
        return max(nums[0], nums[1])
    }
   return max(rob1(nums[:len(nums)-1]), rob1(nums[1:]))
}

func rob1(nums []int) int {
    if len(nums) < 2 {
        return nums[0]
    }
    var dp []int = make([]int, len(nums))
    dp[0] = nums[0]
    dp[1] = max(nums[0], nums[1])
    for i := 2; i < len(nums); i++ {
        dp[i] = max(dp[i-1], dp[i-2] + nums[i])
    }
    return dp[len(nums) - 1]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}